source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+module system correctly throws an error saying that our definition
+doesn't match the options type.
+
+```diff
+diff --git a/default.nix b/default.nix
+index cb423a9..0a9162e 100644
+--- a/default.nix
++++ b/default.nix
+@@ -5,5 +5,9 @@
+       type = lib.types.lines;
+     };
+   };
++
++  config = {
++    generate.script = 42;
++  };
+   
+ }
+```
+
+## Successful evaluation: Assigning a string to generate.script
+
+We can make type checking pass by assigning a string to the option,
+giving us our first successful evaluation of the `generate.script`
+option.
+
+In this case, we assign a script which calls the Google Maps Static API
+to generate a world map, then displaying the result using icat
+(image-cat), both of which are helper scripts.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 0a9162e..24f9c34 100644
+--- a/default.nix
++++ b/default.nix
+@@ -7,7 +7,9 @@
+   };
+ 
+   config = {
+-    generate.script = 42;
++    generate.script = ''
++      map size=640x640 scale=2 | icat
++    '';
+   };
+   
+ }
+```
+
+TODO: Create derivations to get these commands
+
+## A new list option: Declaring generate.requestParams
+
+Let's introduce another option, generate.requestParams. For its type,
+we'll use `listOf <nestedType>`, which is a generic list type where each
+element has to match the given nested type. In our case we want `str` to
+be the nested type, which is a generic string type.
+
+Note that the difference between `str` and `lines` is in their merging
+behavior:
+- For `lines`, multiple definitions get merged by concatenation with
+  newlines
+- For `str`, multiple definitions are not allowed. Which in this case is
+  mostly irrelevant however, since we can't really define a list element
+  multiple times.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 24f9c34..fd5027a 100644
+--- a/default.nix
++++ b/default.nix
+@@ -4,12 +4,21 @@
+     generate.script = lib.mkOption {
+       type = lib.types.lines;
+     };
++
++    generate.requestParams = lib.mkOption {
++      type = lib.types.listOf lib.types.str;
++    };
+   };
+ 
+   config = {
+     generate.script = ''
+       map size=640x640 scale=2 | icat
+     '';
++
++    generate.requestParams = [
++      "size=640x640"
++      "scale=2"
++    ];
+   };
+   
+ }
+```
+
+## Dependencies between options: Using generate.requestParams
+
+A collection of modules generally only has a single option that is meant
+to be evaluated. That's the option that generates the final result we're
+interested in, which in our case is `generate.script`.
+
+In order to build up abstractions on that, we have the ability for
+options to depend on other options. In this case we want the
+`generate.script` option to use the values of `generate.requestParams`.
+We can access the values of options by adding the `config` argument to
+the argument list at the top and using e.g.
+`config.generate.requestParams` to access that options value.
```

Comment:
You need to explain why this works the way that it does. This is the thought process of a new user in this scenario:
- I've defined `generate.script` and `generate.requestParams` next to each other in the `config` attribute.
- If they're next to each other why can't I refer to them? Can't I use `rec`?
- The correct way is to add a function argument named `config`, but I'm currently defining `config`. How on earth does that make sense?